pyccel/codegen/printing/ccode.py

```diff
@@ -1033,15 +1036,21 @@ def _print_Constant(self, expr):
         return self._print(val)
 
     def _print_Return(self, expr):
-        code = ''
         args = [VariableAddress(a) if self.stored_in_c_pointer(a) else a for a in expr.expr]
-        if expr.stmt:
-            code += self._print(expr.stmt)+'\n'
+        if len(args) > 1:
+            if expr.stmt:
+                return self._print(expr.stmt)+'\n'+'return 0;'
+            return 'return 0;'
         if len(args) == 1:
-            code +='return {0};'.format(self._print(args[0]))
-        elif len(args) > 1:
-            code += 'return 0;'
-        return code
+            if isinstance(expr.stmt, CodeBlock):
+                for a in expr.stmt.body:
+                    if isinstance(a, Assign):
+                        b = a
+                if isinstance(b.rhs, IndexedElement):
+                    return '{0}\nreturn {1};'.format(self._print(expr.stmt), self._print(b.lhs))
+                return 'return {0};'.format(self._print(b.rhs))
+            return 'return {0};'.format(self._print(args[0]))
```

Comment:
In a complex case like the one above, we may need to print some of the `expr.stmt`. The only part which we know we can remove is the lhs of the `Assign`

Reply:
By this I mean the case where a one line statement in python is translated to multiple nodes (I can't think of an example which works currently, but there could easily be one in the future). E.g:
```
return [i for i in range(4)]
```
would probably be translated to something like:
```
Allocate(Dummy_0001, 4)
For(Dummy_0002, range(4), Assign(Dummy_001[Dummy_0002], Dummy_0002))
Assign(Out_0001, Dummy_0001)
```
(This is probably not an accurate translation but I can't think of a better example right now) In this case you cannot just print `return Dummy_0001`

Reply:
For example, consider the example given above with multiple `Deallocate` statements. The `Assign` has nothing to do with these `Deallocate` statements. So we can print `'return {0};'.format(self._print(assign_stmt.rhs))`, however this doesn't mean that we can avoid printing all the rest of `expr.stmt`

Reply:
> @EmilyBourne do you mean that we need to print just the deallocation of the array that we returned from.

No!!! Exactly the opposite. We need to make sure that we don't lose any `Deallocate` statements otherwise we will have a memory leak.

Maybe @saniQQ can help you understand this if you are not sure why the `Deallocate` statements are important or why they are in the `Return` object